Fixed-length array accesses still not quite right

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: windows, Microsoft Windows 10 Pro v19044 64-bit
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz  

CC version: cc (Rev3, Built by MSYS2 project) 12.1.0

getwd: C:\Users\[redacted]\Documents\v\acta
vmodules: C:\Users\[redacted]\.vmodules
vroot: C:\v
vexe: C:\v\v.exe
vexe mtime: 2022-07-24 15:25:34
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 2d7406a

Git version: git version 2.37.1
Git vroot status: weekly.2022.28-85-g2d7406a8-dirty
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 125e21e2

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg src/assembler2.v`
See attached file `src/assembler2.v`

**What did you expect to see?**

Successful compilation or a useful error message


**What did you see instead?**
```
==================
C:/Users/[redacted]/AppData/Local/Temp/v_0/assembler2.16017823339585623325.tmp.c:17515: warning: assignment from incompatible pointer type
C:/Users/[redacted]/AppData/Local/Temp/v_0/assembler2.16017823339585623325.tmp.c:18044: error: character array initializer must be a literal, optionally enclosed in braces
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

[assembler2.zip](https://github.com/vlang/v/files/9331583/assembler2.zip)
After #15399 was fixed, I tried retyping the `constant_pool` member of the `Parser` struct as `[][2]string` instead of `[][]string` and fixing the writes to it, but that just exposed another bug. As far as I can tell, the bug is in the handling of line 429, `keys := p.constant_pool.map(it[0])`. It seems like the array of arrays of strings isn't quite getting accesses handled properly.

--- a/vlib/v/gen/c/array.v
+++ b/vlib/v/gen/c/array.v
@@ -341,12 +341,7 @@ fn (mut g Gen) gen_array_map(node ast.CallExpr) {
 	i := g.new_tmp_var()
 	g.writeln('for (int $i = 0; $i < ${tmp}_len; ++$i) {')
 	g.indent++
-	if g.table.sym(inp_info.elem_type).kind == .array_fixed {
-		g.writeln('$inp_elem_type it;')
-		g.writeln('memcpy(&it, (($inp_elem_type*) ${tmp}_orig.data)[$i], sizeof($inp_elem_type));')
-	} else {
-		g.writeln('$inp_elem_type it = (($inp_elem_type*) ${tmp}_orig.data)[$i];')
-	}
+	g.write_prepared_it(inp_info, inp_elem_type, tmp, i)
 	g.set_current_pos_as_last_stmt_pos()
 	mut is_embed_map_filter := false
 	mut expr := node.args[0].expr
@@ -541,7 +536,7 @@ fn (mut g Gen) gen_array_filter(node ast.CallExpr) {
 	i := g.new_tmp_var()
 	g.writeln('for (int $i = 0; $i < ${tmp}_len; ++$i) {')
 	g.indent++
-	g.writeln('$elem_type_str it = (($elem_type_str*) ${tmp}_orig.data)[$i];')
+	g.write_prepared_it(info, elem_type_str, tmp, i)
 	g.set_current_pos_as_last_stmt_pos()
 	mut is_embed_map_filter := false
 	mut expr := node.args[0].expr
@@ -898,7 +893,7 @@ fn (mut g Gen) gen_array_any(node ast.CallExpr) {
 	i := g.new_tmp_var()
 	g.writeln('for (int $i = 0; $i < ${tmp}_len; ++$i) {')
 	g.indent++
-	g.writeln('$elem_type_str it = (($elem_type_str*) ${tmp}_orig.data)[$i];')
+	g.write_prepared_it(info, elem_type_str, tmp, i)
 	g.set_current_pos_as_last_stmt_pos()
 	mut is_embed_map_filter := false
 	mut expr := node.args[0].expr
@@ -972,7 +967,7 @@ fn (mut g Gen) gen_array_all(node ast.CallExpr) {
 	i := g.new_tmp_var()
 	g.writeln('for (int $i = 0; $i < ${tmp}_len; ++$i) {')
 	g.indent++
-	g.writeln('$elem_type_str it = (($elem_type_str*) ${tmp}_orig.data)[$i];')
+	g.write_prepared_it(info, elem_type_str, tmp, i)
 	g.empty_line = true
 	g.set_current_pos_as_last_stmt_pos()
 	mut is_embed_map_filter := false
@@ -1059,4 +1054,13 @@ fn (mut g Gen) write_prepared_tmp_value(tmp string, node &ast.CallExpr, tmp_styp
 	g.writeln(';')
 	g.writeln('int ${tmp}_len = ${tmp}_orig.len;')
 	return has_infix_left_var_name
+}
+
+fn (mut g Gen) write_prepared_it(inp_info ast.Array, inp_elem_type string, tmp string, i string) {
+	if g.table.sym(inp_info.elem_type).kind == .array_fixed {
+		g.writeln('$inp_elem_type it;')
+		g.writeln('memcpy(&it, (($inp_elem_type*) ${tmp}_orig.data)[$i], sizeof($inp_elem_type));')
+	} else {
+		g.writeln('$inp_elem_type it = (($inp_elem_type*) ${tmp}_orig.data)[$i];')
+	}
 }

--- a/vlib/v/gen/c/array.v
+++ b/vlib/v/gen/c/array.v
@@ -341,7 +341,12 @@ fn (mut g Gen) gen_array_map(node ast.CallExpr) {
 	i := g.new_tmp_var()
 	g.writeln('for (int $i = 0; $i < ${tmp}_len; ++$i) {')
 	g.indent++
-	g.writeln('$inp_elem_type it = (($inp_elem_type*) ${tmp}_orig.data)[$i];')
+	if g.table.sym(inp_info.elem_type).kind == .array_fixed {
+		g.writeln('$inp_elem_type it;')
+		g.writeln('memcpy(&it, (($inp_elem_type*) ${tmp}_orig.data)[$i], sizeof($inp_elem_type));')
+	} else {
+		g.writeln('$inp_elem_type it = (($inp_elem_type*) ${tmp}_orig.data)[$i];')
+	}
 	g.set_current_pos_as_last_stmt_pos()
 	mut is_embed_map_filter := false
 	mut expr := node.args[0].expr

--- a/vlib/v/tests/array_of_fixed_array_map_test.v
+++ b/vlib/v/tests/array_of_fixed_array_map_test.v
@@ -1,0 +1,12 @@
+fn test_array_of_fixed_array_map() {
+	mut db := [][2]string{}
+
+	db << ['aaa', '111']!
+	db << ['bbb', '222']!
+	db << ['ccc', '333']!
+
+	keys := db.map(it[0])
+
+	println(keys)
+	assert keys == ['aaa', 'bbb', 'ccc']
+}

--- a/vlib/v/tests/array_of_fixed_array_map_test.v
+++ b/vlib/v/tests/array_of_fixed_array_map_test.v
@@ -1,4 +1,4 @@
-fn test_array_of_fixed_array_map() {
+fn test_array_of_fixed_array_map_filter_any_all() {
 	mut db := [][2]string{}
 
 	db << ['aaa', '111']!
@@ -9,4 +9,9 @@ fn test_array_of_fixed_array_map() {
 
 	println(keys)
 	assert keys == ['aaa', 'bbb', 'ccc']
+
+	assert db.map(it[0] == 'aaa') == [true, false, false]
+	assert db.filter(it[0] == 'bbb') == [['bbb', '222']!]
+	assert db.any(it[0] == 'aaa') == true
+	assert db.all(it[0] == 'aaa') == false
 }

